### PR TITLE
Connect watcher to StreamRegistry for subscriptions

### PR DESCRIPTION
**Pull request by jtakalai**
This maybe could be kept in a separate branch ("brubeck" or something) because it's supposed to run on alongside the current `master` version

### CODE REVIEW

**Review by samt1803 — APPROVED** [review]
LGTM!

File: src/CoreAPIClient.ts [hunk 1]
```diff
@@ -81,4 +86,22 @@ export default class CoreAPIClient {
                 })
             })
     }
+
+    private async _get(apiUrl: string): Promise<any> {
```
Comment (kare) [inline src/CoreAPIClient.ts:90]: Methods for POST and GET should return the same type to make the API uniform and easy to understand.

File: src/main.ts [hunk 2]
```diff
@@ -1,11 +1,25 @@
...
 import Watcher from "./watcher"
 import CoreAPIClient from "./CoreAPIClient"
+
 import MarketplaceJSON from "../lib/marketplace-contracts/build/contracts/Marketplace.json"
+import StreamRegistryJSON from "../lib/streamregistry/StreamRegistryV3.json"
+
+import { TransactionReceipt } from "ethers/providers"
+
+type EthereumAddress = string
+type Permission = {
```
Comment (kare) [inline src/main.ts:16]: Why these are type aliases? Should't they be interfaces or classes to allow extensibility?

File: src/main.ts [hunk 3]
```diff
@@ -43,7 +59,7 @@ async function main(): Promise<void> {
         log.error(`Expected a valid Ethereum key for environment variable ${DEVOPS_KEY}="${devopsKey}".`)
         process.exit(1)
     }
-    let provider = null
+    let provider: ethers.providers.Provider
```
Comment (kare) [inline src/main.ts:62]: 👍

File: src/main.ts [hunk 4]
```diff
@@ -85,56 +107,91 @@ async function main(): Promise<void> {
...
+            endsAt: string,
+    }) => {
+        const {
+            blockNumber,
+            blockIndex,
+            product: productId,
+            address,
+            endsAt,
+        } = args
+        log.info("Watcher > subscribed event at block %s index %s: until %s", blockNumber, blockIndex, endsAt)
```
Comment (kare) [inline src/main.ts:125]: These log messages are written from `main.ts` not `watchers.ts`. Looks like the prefix `Watcher > ...` is a bit misleading?

**Review by kare — CHANGES_REQUESTED** [review]
Looks like a great start, but some refactoring is required.

- [ ] See question on line 137. Why return types have changed for `watcher.on()`?
- [ ] Why no tests or updates to test files?
- [ ] README needs an update

File: src/main.ts [hunk 5]
```diff
@@ -85,56 +107,91 @@ async function main(): Promise<void> {
...
     // write on disk how many blocks have been processed
     const store = new LastBlockStore(lastBlockDir)
     if (process.env["CI"]) {
         store.write(12340000)
     }
-    await watcher.on("eventSuccessfullyProcessed", async (...args: any[]): Promise<any> => {
-        const event = args[0]
+    await watcher.on("eventSuccessfullyProcessed", async (event: any) => {
         store.write(event.blockNumber.toString())
-        return Promise.resolve()
```
[the remark this answers is not in the document]
Comment (jtakalai) [reply src/main.ts:137]: async function returns a Promise as-is, no need to explicitly create it. When execution runs off the end of the async function, the promise is resolved (with void type). Use on `any` is bad there, it should be just `void`.

**Pull request merged**